packages/flutter_tools/lib/src/android/deferred_components_setup_validator.dart

```diff
@@ -0,0 +1,766 @@
+  /// the needed string resources are in the base module `strings.xml`.
+  /// If not, this method will generate a modified `strings.xml` (or a
+  /// completely new one if the original file did not exist) in the
+  /// validator's output directory.
+  void checkAndroidResourcesStrings(List<DeferredComponent> components) {
+    final Directory androidDir = env.projectDir.childDirectory('android');
+    List<String> lines;
+
+    // Add component name mapping to strings.xml
+    final File stringRes = androidDir
+      .childDirectory('app')
+      .childDirectory('src')
+      .childDirectory('main')
+      .childDirectory('res')
+      .childDirectory('values')
+      .childFile('strings.xml');
+    _inputs.add(stringRes);
+    final File stringResOutput = _outputDir
+      .childDirectory('app')
+      .childDirectory('src')
+      .childDirectory('main')
+      .childDirectory('res')
+      .childDirectory('values')
+      .childFile('strings.xml');
+    final Map<String, String> requiredEntriesMap  = <String, String>{};
+    for (final DeferredComponent component in components) {
+      requiredEntriesMap['${component.name}Name'] = component.name;
+    }
+    if (stringResOutput.existsSync()) {
+      stringResOutput.deleteSync();
+    }
+    // Check if all required lines are present.
+    if (stringRes.existsSync()) {
+      lines = stringRes.readAsLinesSync();
+      final Map<String, String> incorrectEntries = <String, String>{}; // Entries
+      final Map<String, String> incorrectEntriesOriginal = <String, String>{};
+      try {
+        final XmlDocument document = XmlDocument.parse(stringRes.readAsStringSync());
+        for (final XmlElement element in document.findAllElements('string')) {
+          final String name = element.getAttribute('name');
+          if (requiredEntriesMap.containsKey(name)) {
+            final String storedString = element.text;
+            if (storedString != null && storedString != requiredEntriesMap[name]) {
+              incorrectEntries[name] = requiredEntriesMap[name];
+              incorrectEntriesOriginal[name] = storedString;
+            }
+            requiredEntriesMap.remove(name);
+          }
+        }
+      } on XmlParserException {
+        throwToolExit('Error parsing $stringRes '
+                      'Please ensure that the strings resource file is a valid XML document and try again.');
+      }
+      if (requiredEntriesMap.isNotEmpty || incorrectEntries.isNotEmpty) {
+        final List<String> modifiedLines = <String>[];
+        // Modify existing entries
+        if (incorrectEntries.isNotEmpty) {
+          for (int i = 0; i < lines.length; i++) {
+            String line = lines[i];
+            for (final String key in incorrectEntries.keys) {
+              // Matches string resource entries that have the key as the name.
+              final RegExp stringNameMatcher = RegExp(r'[\s\n]*<[\s\n]*string[\s\n]+name[\s\n]*=[\s\n]*"' + key);
```

Comment:
Shouldn't the xml parser give you access to names/properties in a way that doesn't require looking at raw lines?

Reply:
The idea is that I want to maintain the original formatting and comments in the file, and the xml parser consumes all of this and collapses it into a format-less xmldocument/xmlelement. This manual parsing is able to maintain any developer formatting/comments.

Reply:
Moved to xml package based system.